### ISSUE TITLE
ability to render an image of generated file from CLI

### ISSUE THREAD

**Issue by jtoy** [jscad/OpenJSCAD.org#325]
would be wonderful if we could output a jpg/png of the rendered object from the command line.
We would probably need the ability to set camera angles for said rendered item.
if this functionality wasn't added directly to the CLI tool, how would one go about doing this all from the CLI?

**Comment by z3dev** [comment]
@jtoy I'm not sure but there are application frameworks that are available to display graphics. If that is possible then the conversion of the display buffer to PNG/JPG is typically available as well.

Maybe this is a feature for the new application. But that's yet to be determined.

**Comment by kaosat-dev** [comment]
actually, this is *somewhat* trivial to add, with a few possible limtations
- I already did something like this for a few projects
- ZoomToFit allows automatically fitting the shapes in the camera view (this will be present in the new viewer)
- you can render things to a file using the SAME renderer that will be in the viewer using https://github.com/stackgl/headless-gl
- main limitation is , you need to run the CLI on a machine that has opengl support
- I can add this in a few hours :)

**Comment by jtoy** [comment]
Love it, would love to see this in the app. Would we be able to configure the camera angle? I have seen apps that implement their own version of tiny renderer: https://github.com/ssloy/tinyrenderer  in case we don’t want OpenGL, but I think expecting OpenGL is fine, thoughts?

**Comment by kaosat-dev** [comment]
yes, camera angle will be configurable :)
Honestly I tried to dance around the need for OpenGL for around a year (for example for docker containters), never found a solution that really works :( but if you do that would be great :)
In the meantime, OpenGL only is fine I think.

**Comment by jtoy** [comment]
I use a non OpenGL rendered all the time with pybullet. They implemented tiny renderer. I have not implemented it myself before,maybe one day, need to learn more C.

**Comment by jtoy** [comment]
btw, in the meantime, I've added a python pipeline into my own workflow that allows me to render stl  into images, works, but I'd much rather have it all done in OpenJSCAD.

**Comment by jtoy** [comment]
just found this nodejs version to render stls to images: https://www.npmjs.com/package/node-stl-thumbnailer
would it be useful if i integrated this into OpenJSCAD?

**Comment by Vishnu-C** [comment]
@napter Have a look on VTK , one can read STL, set the camera angle and render as images. Also it can be all done in offscreen, i.e without even having to load up a 3D render.

**Comment by z3dev** [comment]
See the post in the User Group. @kaosat-dev added support for rendering JSCAD designs from the command line.

https://plus.google.com/112899217323877236232/posts/FmSPdYC4xHr

**Comment by johnwebbcole** [comment]
I was looking through the archives on https://openjscad.nodebb.com but I didn't find the referred post there.  @kaosat-dev did you have a way to render designs to a PNG or other image?

**Comment by z3dev** [comment]
@johnwebbcole I think that you are long for this.

https://github.com/jscad/OpenJSCAD.org/tree/V2/packages/utils/regl-renderer

**Comment by kaosat-dev** [comment]
@johnwebcole , yes it should work rather well using @z3devs link, this file to be more precise https://github.com/jscad/OpenJSCAD.org/blob/V2/packages/utils/regl-renderer/demo-cli.js

You need to install pngjs manually if you want to try it, had to remove it as a dependency as it does not work on TravisCi (local use should be trivial)

**Comment by johnwebbcole** [comment]
Thx for the directions, I was able to get it working in my unit tests for jscad-utils.

**Comment by kaosat-dev** [comment]
good to know @johnwebbcole :)
(btw I really need to get in touch with you about jscad-utils soonish, lots of very useful things that we would love to have for V2)

**Comment by z3dev** [comment]
@jtoy please try using the new rendered inside your pipeline.

https://github.com/jscad/OpenJSCAD.org/tree/V2/packages/utils/regl-renderer

If that works then this thread can be closed. ;)
